bam_qc_metrics/bam_qc.py

```diff
@@ -15,103 +51,73 @@ class bam_qc:
         'run name',
         'sample'
     ]
-    PRECISION = 1 # number of decimal places for rounded output
-    FINE_PRECISION = 3 # finer precision, eg. for reads_per_start_point output
     RANDOM_SEED = 42
-    READ_1_INDEX = 0
-    READ_2_INDEX = 1
-    READ_UNKNOWN_INDEX = 2
-    START_POINTS_KEY = 'start points'
-
-    def __init__(self, bam_path, target_path, expected_insert_max, metadata_path=None,
-                 mark_duplicates_path=None, n_as_mismatch=False, skip_below_mapq=None,
-                 reference=None, sample_rate=None, tmpdir=None, verbose=True):
-        # define instance variables
-        self.bedtools_metrics = None
-        self.custom_metrics = None
-        self.expected_insert_max = expected_insert_max
-        self.mark_duplicates_metrics = {
-            "ESTIMATED_LIBRARY_SIZE": None,
-            "HISTOGRAM": {},
-            "LIBRARY": None,
-            "PERCENT_DUPLICATION": None,
-            "READ_PAIRS_EXAMINED": None,
-            "READ_PAIR_DUPLICATES": None,
-            "READ_PAIR_OPTICAL_DUPLICATES": None,
-            "UNMAPPED_READS": None,
-            "UNPAIRED_READS_EXAMINED": None,
-            "UNPAIRED_READ_DUPLICATES": None
-        }
-        self.metadata = None
-        self.mismatches_by_read = None
-        self.qc_input_bam_path = None
+
+    def __init__(self, config):
+        self.validate_config_fields(config)
+        # read instance variables from config
+        self.expected_insert_max = config[self.CONFIG_KEY_INSERT_MAX]
+        self.mark_duplicates_metrics = self.read_mark_dup(config[self.CONFIG_KEY_MARK_DUPLICATES])
+        self.metadata = self.read_metadata(config[self.CONFIG_KEY_METADATA])
```

Comment:
`read_metadata` is called before `self.verbose` is defined. 

Leads to this exception:
```
 run_bam_qc.py -b /oicr/data/archive/seqware/seqware_analysis_12/hsqwprod/seqware-results/BwaMem_1.2/23809194/SWID_14733394_DCRT_020_Br_R_PE_243_EX_obs470_P020_190718_M00753_0143_000000000-D6DB8_AGATCGCA_L001_001.annotated.bam -o del.json -t S31285117_Covered.bed
Traceback (most recent call last):
  File "/home/slazic/.local/miniconda3/envs/bamqc/bin/run_bam_qc.py", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/home/slazic/Documents/Code/2019-07-25-bam-qc-metrics/bin/run_bam_qc.py", line 120, in <module>
    main()
  File "/home/slazic/Documents/Code/2019-07-25-bam-qc-metrics/bin/run_bam_qc.py", line 116, in main
    qc = bam_qc(config)
  File "/home/slazic/Documents/Code/2019-07-25-bam-qc-metrics/bam_qc_metrics/bam_qc.py", line 61, in __init__
    self.metadata = self.read_metadata(config[self.CONFIG_KEY_METADATA])
  File "/home/slazic/Documents/Code/2019-07-25-bam-qc-metrics/bam_qc_metrics/bam_qc.py", line 168, in read_metadata
    if self.verbose: sys.stderr.write("Metadata file not given, using empty defaults\n")
AttributeError: 'bam_qc' object has no attribute 'verbose'
```

Reply:
Oh, good catch. I'll fix that.

Reply:
Fixed, along with another minor bug.